Infura Ethereum (Independent Publisher)

sebastianzolg opened the pull request:
---
### When submitting a connector, please make sure that you follow the requirements below, otherwise your PR might be rejected. We want to make you have a well-built connector, a smooth certification experience, and your users are happy :) 

#### Important: Our certification engineer, Srikanth, is currently on emergency leave. Please expect a delay with your PR. He will return January 31st, 2022. We apologize for any inconveniences.

If this is your first time submitting to GitHub and you need some help, please sign up for this [session](https://forms.office.com/pages/responsepage.aspx?id=KtIy2vgLW0SOgZbwvQuRaXDXyCl9DkBHq4A2OG7uLpdUMTFJWFFGVUxBNUFZQjZWRUdaOE5BMFkwNS4u). 

- [x] I attest that the connector works and I verified by deploying and testing all the operations. 
- [X] I attest that I have added detailed descriptions for all operations and parameters in the swagger file.
- [X] I attest that I have added response schemas to my actions, unless the response schema is dynamic. 
- [X] I validated the swagger file, `apiDefinition.swagger.json`, by running `paconn validate` command.
- [X] If this is a certified connector, I confirm that `apiProperties.json` has a valid brand color and doesn't use an invalid brand color, `#007ee5` or `#ffffff`. If this is an independent publisher connector, I confirm that I am not submitting a connector icon.

If you are an Independent Publisher, you must also attest to the following to ensure a smooth publishing process:
- [X] I have named this PR after the pattern of "Connector Name (Independent Publisher)" ex: HubSpot Marketing (Independent Publisher)
- [X] Within this PR markdown file, I have pasted screenshots that show: 3 unique operations (actions/triggers) working within a Flow. This can be in one flow or part of multiple flows. For each one of those flows, I have pasted in screenshots of the Flow succeeding. 
- [X] Within this PR markdown file, I have pasted in a screenshot from the Test operations section within the Custom Connector UI.
- [X] If the connector uses OAuth, I have provided detailed steps on how to create an app in the readme.md. 

## Sample Flow
![image](https://user-images.githubusercontent.com/12109119/151934655-9acd71b2-431e-4253-babe-558b96b1de12.png)
![image](https://user-images.githubusercontent.com/12109119/151934691-46d18ef7-a36f-4425-9969-2ceae1e2c576.png)

## Test Action Results
![image](https://user-images.githubusercontent.com/12109119/151932794-799dda42-955f-4c0a-9ad6-083bedf1b0c4.png)

schabungbam commented:
Few comments--
Given that the connector using script, it would be good to add some doc to describe how it is implemented.

Q: Why do we need to transform the request for Get Balance?
It seems overly complex. The connector should simplify the experience for the end user. It seems we are asking the user to specify an array of object (Address, Block), and the connector is only using the first element of the array. This is what I think the payload looks like:
{
   "jsonrpc": "2.0",
   "method": "eth_getBalance",
   "params": [
      {
         "Address": "stringvalue",
         "Block": "string-enum"
      }
   ]
}

Instead, suggest that we ask the user to specify a simpler format, and handle the complexity in the connector. Suggest we change the input to something simpler like:
{
   "jsonrpc": "2.0",
   "method": "eth_getBalance",
   // "params": {  // optional - if we use this, make this required
         "Address": "stringvalue",   // required
         "Block": "string-enum"     // required, a default value would be nice
   // }
}

schabungbam commented on independent-publisher-connectors/Infura Ethereum/script.csx:
```diff
@@ -0,0 +1,115 @@
+    public async Task TransformRequestAsync()
+    {
+        // Check if the operation ID matches what is specified in the OpenAPI definition of the connector
+        if (IsOperationId("Eth_getBalance"))
```
Like this pattern.

schabungbam commented on independent-publisher-connectors/Infura Ethereum/script.csx:
```diff
@@ -0,0 +1,115 @@
+        // Parse as JSON object
+        var contentAsJson = JObject.Parse(contentAsString);
+
+        JArray parameters = (JArray)contentAsJson["params"];
```
See my comments - there may be an opportunity to simplify this

schabungbam commented on independent-publisher-connectors/Infura Ethereum/README.md:
```diff
@@ -0,0 +1,33 @@
+## Obtaining Credentials
+
+The Infura Ethereum Connector uses the Infura Project Id to communicate with the JSON-RPC endpoint. See the **Getting Started** section on how to get a new Project Id.
+The **Project Id** is considered a secret and securely stored inside the Power Platform connection. Do not share the Project Id.
```
Interesting. The Project Id is supposed to be secret, but it is sent a path parameter.

sebastianzolg replied:
Hello @schabungbam, 

I'd like to have a discussion with you on the Project Id. The Infura API docs state the following:
> By default, the Project Secret is optional. If you know you will only be interacting with Infura APIs from a secure environment (e.g. your backend web application), you can toggle the Private Secret Required setting in your Project's configuration page under the Security section.

I think this discussion is as old as the web, but can we assume that a `securestring` parameter inside the Power Platform is never exposed externally, even when used in the path parameter? In other words, is the platform considered a 'secure environment'? 

Having said that, the Infura API does support basic auth as additional protection. However, since basic auth is just a hashed authorization token, this doesn't add much to the security if we consider the Power Platform a 'client' system instead of a 'secure environment'.

In the effort to make this as secure as possible, I tried to add basic auth to the connector (in addition to the project id), and I failed. The Infura API has a very unfortunate implementation of basic auth and expects an empty username:

`curl --user :YOUR-PROJECT-SECRET https://<network>.infura.io/v3/YOUR-PROJECT-ID`
> Note that the user field of the request is left blank and the project id is passed in the URL.

This scenario seems to be unsupported by custom connectors since the username can't be empty in the UI when setting the connector to use basic auth. 

I tried to tweak the apiProperties.json to make this work, but I've failed. Here's my approach:

```json
"connectionParameters": {
      "username": {
        "type": "string",
        "defaultValue": "",
        "uiDefinition": {
          "displayName": "Username",
          "description": "The Username for this api",
          "tooltip": "Provide the Username",
          "constraints": {
            "tabIndex": 2,
            "clearText": true,
            "required": "false",
            "hidden": "true"
          }
        }
      },
      "password": {
        "type": "securestring",
        "uiDefinition": {
          "displayName": "Project Secret",
          "description": "The Project Secret for this api",
          "tooltip": "Provide the Project Secret",
          "constraints": {
            "tabIndex": 3,
            "clearText": false,
            "required": "true"
          }
        }
      },
      "projectId": {
        "type": "securestring",
        "uiDefinition": {
          "constraints": {
            "required": "true",
            "tabIndex": 4,
            "clearText": false
          },
          "description": "The Id of your Infura Ethereum Project.",
          "displayName": "Project Id",
          "tooltip": "The Project Id can be found on the Infura Dashboard. Select your Ethereum Project and check out the KEYS section."
        }
      }
    }
```
As you can see, I attempted to make the username for basic auth to be non-mandatory and set an empty string as the default value. Unfortunately, while the resulting UI looks promising, it doesn't generate the correct hash for basic auth when testing the connector. So, unfortunately, I don't have any insights into what is happening under the hood.

![image](https://user-images.githubusercontent.com/12109119/152643141-15d60ded-90bf-4be9-93e7-5759eff2518a.png)

Also, the `hidden: "true"` attribute doesn't work, which would improve the UX if the approach overall were functional.

I favor going with just the project id, even when used in the path (assuming the Power Platform to be a 'secure environment'). 

Would you mind having a look? Are you part of the product group, and could this be a feature request or considered a 'known limitation', or should I report this to @natalie-pienkowska?

sebastianzolg replied:
I've added a **Securing Infura** section to the README.md. It should make the solution secure, even without basic auth. Still, I'd love to follow up on the discussion about this scenario.

sebastianzolg commented:
@schabungbam Thank you for your kind feedback! Those are good points. Give me a few days to address them. 
—Sebastian

natalie-pienkowska commented:
Hi @sebastianzolg just wanted to let you know that we have a delay in our certification pipeline due to a family emergency. We will let you know when we can review your PR shortly.

sebastianzolg commented:
Dear @natalie-pienkowska, there is absolutely no need to rush! Please, everyone, take the time to overcome this emergency. There are more important things than a custom connector. I wish you all the best. Take care, everyone!

sebastianzolg commented:
Hello @schabungbam, again, thank you very much for your comments and thoughts. Please see my answer on the **Project Id** [here](https://github.com/microsoft/PowerPlatformConnectors/pull/1410#discussion_r800056212).

Regarding simplifying the `eth_getBalance`, I followed your advice and removed the need to wrap the input in an array. Please note that the Infura API still expects a flat array for this case. I've improved the [script.csx](https://github.com/microsoft/PowerPlatformConnectors/pull/1410/commits/8185b1d1b7a0b1c756da5d67b4511a435c011141) to generalize this approach. Also, I think this approach is easy to maintain and allows me to add more calls in the future without doing too much transformation. The UX is good enough, IMO. I've also added a default value for the `Block` parameter. 

Furthermore, I did improve the [README.md](https://github.com/microsoft/PowerPlatformConnectors/pull/1410/commits/250d27e1917f6ae7003afd3433a89b009582f91b) and added a section to explain transformation code and the route policy.

Let me know what you think, and see what @natalie-pienkowska **+ team** has to say about it once everyone is back to work.

—Sebastian

sriyen-msft commented on independent-publisher-connectors/Infura Ethereum/README.md:
```diff
@@ -0,0 +1,59 @@
+# Infura Ethereum Connector
+
+Infura provides services to access the Ethereum Blockchain without your own Ethereum Node. This connector allows you to interact with the Ethereum Blockchain JSON-RPC API through Infura. Common use cases include checking your Ethereum account's (Wallet) balance, current gas fees, or information on the current block.
+
```
@sebastianzolg Please refer the [template provided for readme.md file](https://github.com/microsoft/PowerPlatformConnectors/blob/dev/templates/Independent%20Publisher/readme.md) and add required sections.

sebastianzolg replied:
Added Publisher Name, Known Issues, and Deployment Instructions.

sriyen-msft commented on independent-publisher-connectors/Infura Ethereum/apiDefinition.swagger.json:
```diff
@@ -0,0 +1,258 @@
+                "jsonrpc": {
+                  "type": "string",
+                  "description": "jsonrpc",
+                  "title": "",
```
@sebastianzolg Empty property values can be removed until unless you have a specific reason to keep them.

sebastianzolg replied:
Removed all empty `title: ""` properties.

sebastianzolg commented:
Hello @sriyen-msft,

Thanks for the review. I've updated the content accordingly. Please have a look and let me know if there is anything else I need to change. 

Thank you.
—Sebastian

sebastianzolg commented:
Hello @sriyen-msft, I hope you do well. Is there any update on this? What are the next steps, and do you need anything else from me? Thank you.

sriyen-msft commented:
[[certify-connector]]

sriyen-msft commented:
@sebastianzolg  We are preparing your connector for production deployment and will queue up for the next deployment schedule. Once your connector onboards the next deployment schedule, it starts to deploy your connector in our production environments which typically takes 3-4 weeks. Please note that the documentation for your connector will not be generated until the deployment process is in the final steps. Thanks.

Amjed-Ayoub commented:
Dear partner,

We apologize for the delay you might be experiencing while we are working on a high volume of connector requests. Rest assured that we are doing everything we can to address this volume. We are taking measures like improving our processes and hiring additional team members to guarantee resolving your connector requests in a timely manner. 

Please let us know if you have any questions in the meantime.

sriyen-msft commented:
@sebastianzolg  Sorry for the delay in response, your connector is now in production deployment and reached to [Asia](https://docs.microsoft.com/en-us/connectors/custom-connectors/certification-submission#region-deployment) region, and it would be completed and available in all the production regions by Friday 11th February 2022. Please let us know if you have any questions, Thanks.

sriyen-msft commented:
@sebastianzolg Congratulations! The deployment of your connector and documentation page is now complete, We have a short [survey](https://forms.office.com/Pages/ResponsePage.aspx?id=v4j5cvGGr0GRqy180BHbR0BHMrjhL0hDmckROosW6AFURERIVUUzMk5XRzJGV1dJM1Q5RVZJUjFXSC4u) on the connector certification process that we would like for you to complete. Your feedback is appreciated and will help us improve the program, Thanks.

sriyen-msft commented:
@sebastianzolg We have found that the deployment of your connector had issues at our end. We have fixed it and redeploying your connector to all the production regions. Please allow us some time to get back to you with ETA for completing the deployment. We will keep you posted with the updates. Apologies for the inconvenience. 
@natalie-pienkowska @Amjed-Ayoub FYI

sriyen-msft commented:
@sebastianzolg We have deployed your connector to Preview and are able to create a connection as per the instructions mentioned in the readme.md and test the connector. Can you please test the connector in the Preview region and let us know if we can go ahead and queue up the connector for production deployment.

Docs - [Create an environment in the Preview region](https://docs.microsoft.com/en-us/connectors/custom-connectors/certification-testing#create-an-environment-in-the-preview-region)

![image](https://user-images.githubusercontent.com/56276394/164343766-32235324-0530-46cf-a130-6ab32ed5a508.png)

sebastianzolg commented:
@sriyen-msft Good day! The connector is working inside a preview environment. Please go ahead and queue up the connector for production deployment. Thanks a lot for your help and support. 
—Sebastian
